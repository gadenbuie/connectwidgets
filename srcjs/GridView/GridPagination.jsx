import { range } from 'lodash';
import PropTypes from 'prop-types';
import './gridPagination.scss';

const btnClass = {
  base: 'rscgrid-pagination__btn page-link',
  current: 'rscgrid-pagination__btn rscgrid-pagination__btn--current page-link',
  disabled: 'rscgrid-pagination__btn rscgrid-pagination__btn--disabled page-link',
};

const bsPageClass = {
  base: 'page-item',
  disabled: 'page-item disabled',
  current: 'page-item active',
}

function GridPagination({ total, perPage, currentPage, onChange }) {
  const sep = '...';
  const totalPages = Math.ceil(total / perPage);

  function genPagingRange() {
    if (totalPages < 7) {
      return range(1, totalPages + 1);
    }
    if (totalPages - currentPage < 3) {
      return [1, sep, ...range(totalPages - 3, totalPages + 1)];
    }
    if (currentPage > 4) {
      return [1, sep, ...range(currentPage - 1, currentPage + 2), sep, totalPages];
    }
    return [...range(1, 6), sep, totalPages];
  }

  function pageButtons() {
    return genPagingRange().map((pageNum, index) => {
      if (pageNum === sep) {
        return (
          <li
            className="rscgrid-pagination__sep"
            key={index}
            role="separator"
          >...</li>
        );
      }
      return (
        <li
          className={currentPage === pageNum ? bsPageClass.current : bsPageClass.base}
          key={index}
        >
          {/* eslint-disable */}
          <a
            className={currentPage === pageNum ? btnClass.current : btnClass.base}
            href="#"
            onClick={(ev) => { ev.preventDefault(); gotoPage(pageNum); }}
          >{pageNum}</a>
          {/* eslint-enable */}
        </li>
      );
    });
  }

  function gotoPage(page) {
    const reachedLimit = page == 0 || page > totalPages;
    if (!reachedLimit && page !== currentPage) {
      onChange(page);
    }
  }

  function disablePrevious() {
    return currentPage === 1;
  }

  function disableNext() {
    return currentPage === totalPages || totalPages === 0;
  }

  return (
    <nav className="rscgrid-pagination">
      <ul className="rscgrid-pagination__list pagination">
        <li
          className={disablePrevious() ? bsPageClass.disabled : bsPageClass.base}
        >
          {/* eslint-disable */}
          <a
            className={disablePrevious() ? btnClass.disabled : btnClass.base}
            href="#"
            onClick={(ev) => { ev.preventDefault(); gotoPage(currentPage - 1); }}
          >❮</a>
          {/* eslint-enable */}
        </li>
        {pageButtons()}
        <li
          className={disableNext() ? bsPageClass.disabled : bsPageClass.base}
        >
          {/* eslint-disable */}
          <a
            className={disableNext() ? btnClass.disabled : btnClass.base}
            href="#"
            onClick={(ev) => { ev.preventDefault(); gotoPage(currentPage + 1); }}
          >❯</a>
          {/* eslint-enable */}
        </li>
      </ul>
    </nav>
  );
}

GridPagination.propTypes = {
  total: PropTypes.number,
  perPage: PropTypes.number,
  currentPage: PropTypes.number,
  onChange: PropTypes.func,
}

export default GridPagination;
