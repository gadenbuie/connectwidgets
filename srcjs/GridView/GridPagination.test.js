import { shallow } from 'enzyme';
import toJson from 'enzyme-to-json';
import GridPagination from './GridPagination';

let wrapper;
const onChangeFn = jest.fn().mockImplementation(page => {
  wrapper.setProps({ currentPage: page });
});

const mkWrapper = (totalItems = 63, currentPage = 1) => {
  wrapper = shallow(<GridPagination
    total={totalItems}
    perPage={12}
    currentPage={currentPage}
    onChange={onChangeFn}
  />);
};

describe('<GridPagination />', () => {
  beforeEach(() => {
    jest.clearAllMocks();
  });

  it('renders properly', () => {
    mkWrapper();
    const pagesBtns = wrapper.find('.rscgrid-pagination__btn');
    expect(toJson(wrapper)).toMatchSnapshot();
    expect(pagesBtns.length).toBe(8);
    expect(pagesBtns.at(1).hasClass('rscgrid-pagination__btn--current')).toBe(true);
  });
  
  it('onChange: navigates and highlights the current page as expected', () => {
    mkWrapper();
    const pagesBtns = () => wrapper.find('.rscgrid-pagination__btn');
    expect(pagesBtns().at(0).hasClass('rscgrid-pagination__btn--disabled')).toBe(true);

    // hit next page
    pagesBtns().at(7).simulate('click', { preventDefault: () => {} });
    expect(onChangeFn).toHaveBeenCalledWith(2);
    expect(pagesBtns().at(0).hasClass('rscgrid-pagination__btn--disabled')).toBe(false);
    expect(pagesBtns().at(2).hasClass('rscgrid-pagination__btn--current')).toBe(true);

    // hit fourth page
    pagesBtns().at(4).simulate('click', { preventDefault: () => {} });
    expect(onChangeFn).toHaveBeenCalledWith(4);
    expect(pagesBtns().at(0).hasClass('rscgrid-pagination__btn--disabled')).toBe(false);
    expect(pagesBtns().at(7).hasClass('rscgrid-pagination__btn--disabled')).toBe(false);
    expect(pagesBtns().at(4).hasClass('rscgrid-pagination__btn--current')).toBe(true);

    // hit previous page
    pagesBtns().at(0).simulate('click', { preventDefault: () => {} });
    expect(onChangeFn).toHaveBeenCalledWith(3);
    expect(pagesBtns().at(0).hasClass('rscgrid-pagination__btn--disabled')).toBe(false);
    expect(pagesBtns().at(7).hasClass('rscgrid-pagination__btn--disabled')).toBe(false);
    expect(pagesBtns().at(3).hasClass('rscgrid-pagination__btn--current')).toBe(true);

    // hit last page
    pagesBtns().at(6).simulate('click', { preventDefault: () => {} });
    expect(onChangeFn).toHaveBeenCalledWith(6);
    expect(pagesBtns().at(0).hasClass('rscgrid-pagination__btn--disabled')).toBe(false);
    expect(pagesBtns().at(7).hasClass('rscgrid-pagination__btn--disabled')).toBe(true);
    expect(pagesBtns().at(6).hasClass('rscgrid-pagination__btn--current')).toBe(true);
  });

  describe('for too many pages', () => {
    it('displays compact pagination controls', () => {
      mkWrapper(1300);
      expect(toJson(wrapper)).toMatchSnapshot();
      expect(wrapper.text()).toBe('❮12345...109❯')

      mkWrapper(1300, 54);
      expect(toJson(wrapper)).toMatchSnapshot();
      expect(wrapper.text()).toBe('❮1...535455...109❯')

      mkWrapper(1300, 6);
      expect(toJson(wrapper)).toMatchSnapshot();
      expect(wrapper.text()).toBe('❮1...567...109❯')

      mkWrapper(1300, 106);
      expect(toJson(wrapper)).toMatchSnapshot();
      expect(wrapper.text()).toBe('❮1...105106107...109❯')
      
      mkWrapper(1300, 109);
      expect(toJson(wrapper)).toMatchSnapshot();
      expect(wrapper.text()).toBe('❮1...106107108109❯')
    });
  });
});
