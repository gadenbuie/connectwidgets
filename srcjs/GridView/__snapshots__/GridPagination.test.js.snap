// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<GridPagination /> for too many pages displays compact pagination controls 1`] = `
<nav
  className="rscgrid-pagination"
>
  <ul
    className="rscgrid-pagination__list pagination"
  >
    <li
      className="page-item disabled"
    >
      <a
        className="rscgrid-pagination__btn rscgrid-pagination__btn--disabled page-link"
        href="#"
        onClick={[Function]}
      >
        ❮
      </a>
    </li>
    <li
      className="page-item active"
      key="0"
    >
      <a
        className="rscgrid-pagination__btn rscgrid-pagination__btn--current page-link"
        href="#"
        onClick={[Function]}
      >
        1
      </a>
    </li>
    <li
      className="page-item"
      key="1"
    >
      <a
        className="rscgrid-pagination__btn page-link"
        href="#"
        onClick={[Function]}
      >
        2
      </a>
    </li>
    <li
      className="page-item"
      key="2"
    >
      <a
        className="rscgrid-pagination__btn page-link"
        href="#"
        onClick={[Function]}
      >
        3
      </a>
    </li>
    <li
      className="page-item"
      key="3"
    >
      <a
        className="rscgrid-pagination__btn page-link"
        href="#"
        onClick={[Function]}
      >
        4
      </a>
    </li>
    <li
      className="page-item"
      key="4"
    >
      <a
        className="rscgrid-pagination__btn page-link"
        href="#"
        onClick={[Function]}
      >
        5
      </a>
    </li>
    <li
      className="rscgrid-pagination__sep"
      key="5"
      role="separator"
    >
      ...
    </li>
    <li
      className="page-item"
      key="6"
    >
      <a
        className="rscgrid-pagination__btn page-link"
        href="#"
        onClick={[Function]}
      >
        109
      </a>
    </li>
    <li
      className="page-item"
    >
      <a
        className="rscgrid-pagination__btn page-link"
        href="#"
        onClick={[Function]}
      >
        ❯
      </a>
    </li>
  </ul>
</nav>
`;

exports[`<GridPagination /> for too many pages displays compact pagination controls 2`] = `
<nav
  className="rscgrid-pagination"
>
  <ul
    className="rscgrid-pagination__list pagination"
  >
    <li
      className="page-item"
    >
      <a
        className="rscgrid-pagination__btn page-link"
        href="#"
        onClick={[Function]}
      >
        ❮
      </a>
    </li>
    <li
      className="page-item"
      key="0"
    >
      <a
        className="rscgrid-pagination__btn page-link"
        href="#"
        onClick={[Function]}
      >
        1
      </a>
    </li>
    <li
      className="rscgrid-pagination__sep"
      key="1"
      role="separator"
    >
      ...
    </li>
    <li
      className="page-item"
      key="2"
    >
      <a
        className="rscgrid-pagination__btn page-link"
        href="#"
        onClick={[Function]}
      >
        53
      </a>
    </li>
    <li
      className="page-item active"
      key="3"
    >
      <a
        className="rscgrid-pagination__btn rscgrid-pagination__btn--current page-link"
        href="#"
        onClick={[Function]}
      >
        54
      </a>
    </li>
    <li
      className="page-item"
      key="4"
    >
      <a
        className="rscgrid-pagination__btn page-link"
        href="#"
        onClick={[Function]}
      >
        55
      </a>
    </li>
    <li
      className="rscgrid-pagination__sep"
      key="5"
      role="separator"
    >
      ...
    </li>
    <li
      className="page-item"
      key="6"
    >
      <a
        className="rscgrid-pagination__btn page-link"
        href="#"
        onClick={[Function]}
      >
        109
      </a>
    </li>
    <li
      className="page-item"
    >
      <a
        className="rscgrid-pagination__btn page-link"
        href="#"
        onClick={[Function]}
      >
        ❯
      </a>
    </li>
  </ul>
</nav>
`;

exports[`<GridPagination /> for too many pages displays compact pagination controls 3`] = `
<nav
  className="rscgrid-pagination"
>
  <ul
    className="rscgrid-pagination__list pagination"
  >
    <li
      className="page-item"
    >
      <a
        className="rscgrid-pagination__btn page-link"
        href="#"
        onClick={[Function]}
      >
        ❮
      </a>
    </li>
    <li
      className="page-item"
      key="0"
    >
      <a
        className="rscgrid-pagination__btn page-link"
        href="#"
        onClick={[Function]}
      >
        1
      </a>
    </li>
    <li
      className="rscgrid-pagination__sep"
      key="1"
      role="separator"
    >
      ...
    </li>
    <li
      className="page-item"
      key="2"
    >
      <a
        className="rscgrid-pagination__btn page-link"
        href="#"
        onClick={[Function]}
      >
        5
      </a>
    </li>
    <li
      className="page-item active"
      key="3"
    >
      <a
        className="rscgrid-pagination__btn rscgrid-pagination__btn--current page-link"
        href="#"
        onClick={[Function]}
      >
        6
      </a>
    </li>
    <li
      className="page-item"
      key="4"
    >
      <a
        className="rscgrid-pagination__btn page-link"
        href="#"
        onClick={[Function]}
      >
        7
      </a>
    </li>
    <li
      className="rscgrid-pagination__sep"
      key="5"
      role="separator"
    >
      ...
    </li>
    <li
      className="page-item"
      key="6"
    >
      <a
        className="rscgrid-pagination__btn page-link"
        href="#"
        onClick={[Function]}
      >
        109
      </a>
    </li>
    <li
      className="page-item"
    >
      <a
        className="rscgrid-pagination__btn page-link"
        href="#"
        onClick={[Function]}
      >
        ❯
      </a>
    </li>
  </ul>
</nav>
`;

exports[`<GridPagination /> for too many pages displays compact pagination controls 4`] = `
<nav
  className="rscgrid-pagination"
>
  <ul
    className="rscgrid-pagination__list pagination"
  >
    <li
      className="page-item"
    >
      <a
        className="rscgrid-pagination__btn page-link"
        href="#"
        onClick={[Function]}
      >
        ❮
      </a>
    </li>
    <li
      className="page-item"
      key="0"
    >
      <a
        className="rscgrid-pagination__btn page-link"
        href="#"
        onClick={[Function]}
      >
        1
      </a>
    </li>
    <li
      className="rscgrid-pagination__sep"
      key="1"
      role="separator"
    >
      ...
    </li>
    <li
      className="page-item"
      key="2"
    >
      <a
        className="rscgrid-pagination__btn page-link"
        href="#"
        onClick={[Function]}
      >
        105
      </a>
    </li>
    <li
      className="page-item active"
      key="3"
    >
      <a
        className="rscgrid-pagination__btn rscgrid-pagination__btn--current page-link"
        href="#"
        onClick={[Function]}
      >
        106
      </a>
    </li>
    <li
      className="page-item"
      key="4"
    >
      <a
        className="rscgrid-pagination__btn page-link"
        href="#"
        onClick={[Function]}
      >
        107
      </a>
    </li>
    <li
      className="rscgrid-pagination__sep"
      key="5"
      role="separator"
    >
      ...
    </li>
    <li
      className="page-item"
      key="6"
    >
      <a
        className="rscgrid-pagination__btn page-link"
        href="#"
        onClick={[Function]}
      >
        109
      </a>
    </li>
    <li
      className="page-item"
    >
      <a
        className="rscgrid-pagination__btn page-link"
        href="#"
        onClick={[Function]}
      >
        ❯
      </a>
    </li>
  </ul>
</nav>
`;

exports[`<GridPagination /> for too many pages displays compact pagination controls 5`] = `
<nav
  className="rscgrid-pagination"
>
  <ul
    className="rscgrid-pagination__list pagination"
  >
    <li
      className="page-item"
    >
      <a
        className="rscgrid-pagination__btn page-link"
        href="#"
        onClick={[Function]}
      >
        ❮
      </a>
    </li>
    <li
      className="page-item"
      key="0"
    >
      <a
        className="rscgrid-pagination__btn page-link"
        href="#"
        onClick={[Function]}
      >
        1
      </a>
    </li>
    <li
      className="rscgrid-pagination__sep"
      key="1"
      role="separator"
    >
      ...
    </li>
    <li
      className="page-item"
      key="2"
    >
      <a
        className="rscgrid-pagination__btn page-link"
        href="#"
        onClick={[Function]}
      >
        106
      </a>
    </li>
    <li
      className="page-item"
      key="3"
    >
      <a
        className="rscgrid-pagination__btn page-link"
        href="#"
        onClick={[Function]}
      >
        107
      </a>
    </li>
    <li
      className="page-item"
      key="4"
    >
      <a
        className="rscgrid-pagination__btn page-link"
        href="#"
        onClick={[Function]}
      >
        108
      </a>
    </li>
    <li
      className="page-item active"
      key="5"
    >
      <a
        className="rscgrid-pagination__btn rscgrid-pagination__btn--current page-link"
        href="#"
        onClick={[Function]}
      >
        109
      </a>
    </li>
    <li
      className="page-item disabled"
    >
      <a
        className="rscgrid-pagination__btn rscgrid-pagination__btn--disabled page-link"
        href="#"
        onClick={[Function]}
      >
        ❯
      </a>
    </li>
  </ul>
</nav>
`;

exports[`<GridPagination /> renders properly 1`] = `
<nav
  className="rscgrid-pagination"
>
  <ul
    className="rscgrid-pagination__list pagination"
  >
    <li
      className="page-item disabled"
    >
      <a
        className="rscgrid-pagination__btn rscgrid-pagination__btn--disabled page-link"
        href="#"
        onClick={[Function]}
      >
        ❮
      </a>
    </li>
    <li
      className="page-item active"
      key="0"
    >
      <a
        className="rscgrid-pagination__btn rscgrid-pagination__btn--current page-link"
        href="#"
        onClick={[Function]}
      >
        1
      </a>
    </li>
    <li
      className="page-item"
      key="1"
    >
      <a
        className="rscgrid-pagination__btn page-link"
        href="#"
        onClick={[Function]}
      >
        2
      </a>
    </li>
    <li
      className="page-item"
      key="2"
    >
      <a
        className="rscgrid-pagination__btn page-link"
        href="#"
        onClick={[Function]}
      >
        3
      </a>
    </li>
    <li
      className="page-item"
      key="3"
    >
      <a
        className="rscgrid-pagination__btn page-link"
        href="#"
        onClick={[Function]}
      >
        4
      </a>
    </li>
    <li
      className="page-item"
      key="4"
    >
      <a
        className="rscgrid-pagination__btn page-link"
        href="#"
        onClick={[Function]}
      >
        5
      </a>
    </li>
    <li
      className="page-item"
      key="5"
    >
      <a
        className="rscgrid-pagination__btn page-link"
        href="#"
        onClick={[Function]}
      >
        6
      </a>
    </li>
    <li
      className="page-item"
    >
      <a
        className="rscgrid-pagination__btn page-link"
        href="#"
        onClick={[Function]}
      >
        ❯
      </a>
    </li>
  </ul>
</nav>
`;
